<!DOCTYPE html>

<html class="writer-html5" lang="en">
<head>
<meta charset="utf-8"/>
<meta content="IE=edge" http-equiv="X-UA-Compatible"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<link href="../img/favicon.ico" rel="shortcut icon"/>
<title>Tutorials - MetaPuf</title>
<link href="../css/theme.css" rel="stylesheet"/>
<link href="../css/theme_extra.css" rel="stylesheet"/>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" rel="stylesheet"/>
<script>
        // Current page data
        var mkdocs_page_name = "Tutorials";
        var mkdocs_page_input_path = "Tutorial.md";
        var mkdocs_page_url = null;
      </script>
<script defer="" src="../js/jquery-3.6.0.min.js"></script>
<!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/languages/yaml.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/languages/rust.min.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
<link href="../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><style>html.glightbox-open { overflow: initial; height: 100%; }</style><script src="../assets/javascripts/glightbox.min.js"></script></head>
<body class="wy-body-for-nav" role="document">
<div class="wy-grid-for-nav">
<nav class="wy-nav-side stickynav" data-toggle="wy-nav-shift">
<div class="wy-side-scroll">
<div class="wy-side-nav-search">
<a class="icon icon-home" href=".."> MetaPuf
        </a>
</div>
<div aria-label="Navigation menu" class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation">
<ul>
<li class="toctree-l1"><a class="reference internal" href="..">Welcome to MetaPUF</a>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../Overview/">Overview</a>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../Workflow/">Workflow</a>
</li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal current" href="./">Tutorials</a>
<ul class="current">
<li class="toctree-l2"><a class="reference internal" href="#installation">Installation</a>
<ul>
<li class="toctree-l3"><a class="reference internal" href="#from-bioconda">From Bioconda</a>
<ul>
<li class="toctree-l4"><a class="reference internal" href="#install-conda-executable">Install conda executable</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#add-bioconda-channels">Add bioconda channels</a>
</li>
<li class="toctree-l4"><a class="reference internal" href="#create-a-conda-environment">Create a conda environment</a>
</li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#from-github-source-code">From Github source code</a>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#inputs">Inputs</a>
</li>
<li class="toctree-l2"><a class="reference internal" href="#quick-start">Quick start</a>
</li>
<li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a>
</li>
</ul>
</li>
</ul>
</div>
</div>
</nav>
<section class="wy-nav-content-wrap" data-toggle="wy-nav-shift">
<nav aria-label="Mobile navigation menu" class="wy-nav-top" role="navigation">
<i class="fa fa-bars" data-toggle="wy-nav-top"></i>
<a href="..">MetaPuf</a>
</nav>
<div class="wy-nav-content">
<div class="rst-content"><div aria-label="breadcrumbs navigation" role="navigation">
<ul class="wy-breadcrumbs">
<li><a alt="Docs" class="icon icon-home" href=".."></a> »</li>
<li>Tutorials</li>
<li class="wy-breadcrumbs-aside">
</li>
</ul>
<hr/>
</div>
<div class="document" itemscope="itemscope" itemtype="http://schema.org/Article" role="main">
<div class="section" itemprop="articleBody">
<h1 id="tutorials">Tutorials</h1>
<p>Here below you can find tutorials on how to use some of the MetaPUF pipelines and standalones. This gives you a flavor of what can be achieved using MetaPUF.</p>
<h2 id="installation"><strong>Installation</strong></h2>
<h3 id="from-bioconda">From Bioconda</h3>
<p>The pipelines in MetaPUF rely on third party software such as Snakemake or mg-toolkit that are not python libraries. We therefore recommend to use conda that provides pre-compiled software for you.</p>
<h4 id="install-conda-executable">Install conda executable</h4>
<p>We recommend to install Miniconda Python3 distribution from <a href="https://conda.io/en/latest/miniconda.html">Download</a>.</p>
<h4 id="add-bioconda-channels">Add bioconda channels</h4>
<p>When you want to install a ny package, conda looks on the official Anaconda website (channel). However many channels are present, we will use bioconda channel. To use this type the command:</p>
<pre><code>conda config --add channels defaults
conda config --add channels bioconda
</code></pre>
<p>It is important to have the channels in the right order. You can check the <a href="https://bioconda.github.io/">order</a> with the command</p>
<pre><code>conda config --get channels
</code></pre>
<h4 id="create-a-conda-environment">Create a conda environment</h4>
<p>Once miniconda is installed and the channels set, you can create a new environment for MetaPUF using the command</p>
<pre><code>conda env --name metapuf create --f metapuf_env.yml
</code></pre>
<p>You must active the metapuf environment whenever you work with this workflow.</p>
<pre><code>conda activate metapuf
</code></pre>
<h3 id="from-github-source-code">From Github source code</h3>
<p>You can also install metapuf locally from Github</p>
<pre><code>conda env --name metapuf create --f metapuf_env.yml
conda activate metapuf
git clone https://github.com/PRIDE-reanalysis/MetaPUF.git
cd MetaPUF
</code></pre>
<h2 id="inputs"><strong>Inputs</strong></h2>
<p>The workflow requires the following inputs in the config.yaml file:</p>
<ul>
<li><code>Study: (ERP124921) | input_dir: Absolute path to the input directory</code></li>
</ul>
<p>This parameter takes in European Nucleotide Archive (ENA) secondary study accession: starts with (ERP|DRP|SRP) followed by six digits.</p>
<p>You can also input your own metagenomics and/or metatranscriptomics assemblies and metaproteomics data from the same samples. To be compatibale with the MetaPUF analysis, there are some instructions</p>
<ul>
<li>
<p>The workflow accepts predicted coding sequence (CDS) from Prodigal (v.2.6.3).</p>
</li>
<li>
<p>The naming convention for the assembly files and the predicted CDS files should comply with the following format:</p>
<ul>
<li>Assembly filename should be defined as assemblyname_FASTA.fasta.gz.</li>
<li>The protein filename should be defined as assemblyname_FASTA_predicted_cds.faa.gz.</li>
<li>For example ERZ4290841_FASTA.fasta.gz where ERZ4290841is the assemblyname.</li>
</ul>
</li>
<li>
<p><code>Pride_id: PRIDE dataset accession For example (PXD003791)</code></p>
</li>
<li>
<p><code>Version: The version of MGnidy analysis.</code></p>
</li>
</ul>
<p>This accepts either "4.1" or "5.0" as string values. The default is "5.0".</p>
<ul>
<li><code>Db_size: Size of the proteins sequence database in bytes.</code></li>
</ul>
<p>The default is 1073741824 (integer data type).</p>
<ul>
<li><code>outputdir: Directory path to save the output</code></li>
</ul>
<p>The user needs to enter a .csv file that contains all the sample information as shown in the example below</p>
<table>
<thead>
<tr>
<th>Sample</th>
<th>PRIDE Accession</th>
<th>Raw file</th>
<th>Raw file URLs</th>
<th>Sample Accession</th>
<th>Assembly</th>
</tr>
</thead>
<tbody>
<tr>
<td>S6</td>
<td>PXD005780</td>
<td>S6.raw</td>
<td>https://ftp.pride.ebi.ac.uk/pride/data/archive/2017/07/PXD005780/S6.raw</td>
<td>ERS1509315</td>
<td>ERZ1669330</td>
</tr>
</tbody>
</table>
<p>This workflow looks for this file in the config directory. You don't need to provide the <code>PRIDE Accession</code> number and <code>Raw file URLs</code>if you have the data locally, and just leave the column as blank, however the header is still needed.</p>
<h2 id="quick-start"><strong>Quick start</strong></h2>
<p>Snakemake dry run command</p>
<pre><code>Snakemake -n --cores 4
</code></pre>
<p>To run the workflow with Snakemake</p>
<pre><code>Snakemake --cores 4
</code></pre>
<h2 id="outputs"><strong>Outputs</strong></h2>
<p>The outputs include the following files
* <code>Protein sequence database</code></p>
<ul>
<li>
<p><code>Report informing assemblies in each protein sequence db</code></p>
</li>
<li>
<p><code>Protein reports from PeptideShaker</code></p>
</li>
<li>
<p><code>Peptide report from PeptideShaker</code></p>
</li>
<li>
<p><code>Post-processing reports</code></p>
</li>
<li>
<p><code>High confidence GFF format report</code></p>
</li>
<li><code>Low confidence GFF format report</code></li>
</ul>
<p><a class="glightbox" href="../dag.png"><img alt="alt text" src="../dag.png" title="Snakemake pipeline"/></a></p>
</div>
</div><footer>
<div aria-label="Footer Navigation" class="rst-footer-buttons" role="navigation">
<a class="btn btn-neutral float-left" href="../Workflow/" title="Workflow"><span class="icon icon-circle-arrow-left"></span> Previous</a>
</div>
<hr/>
<div role="contentinfo">
<!-- Copyright etc -->
</div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
</div>
</div>
</section>
</div>
<div aria-label="Versions" class="rst-versions" role="note">
<span class="rst-current-version" data-toggle="rst-current-version">
<span><a href="../Workflow/" style="color: #fcfcfc">« Previous</a></span>
</span>
</div>
<script>var base_url = '..';</script>
<script defer="" src="../js/theme_extra.js"></script>
<script defer="" src="../js/theme.js"></script>
<script defer="">
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>
<script>const lightbox = GLightbox({"touchNavigation": true, "loop": false, "width": "100%", "height": "auto", "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom"});</script></body>
</html>
